quinn/src/connection.rs

```diff
@@ -764,35 +775,47 @@ impl ConnectionRef {
 
 impl Clone for ConnectionRef {
     fn clone(&self) -> Self {
-        self.lock("clone").ref_count += 1;
+        self.state.lock("clone").ref_count += 1;
         Self(self.0.clone())
     }
 }
 
 impl Drop for ConnectionRef {
     fn drop(&mut self) {
-        let conn = &mut *self.lock("drop");
+        let conn = &mut *self.state.lock("drop");
         if let Some(x) = conn.ref_count.checked_sub(1) {
             conn.ref_count = x;
             if x == 0 && !conn.inner.is_closed() {
                 // If the driver is alive, it's just it and us, so we'd better shut it down. If it's
                 // not, we can't do any harm. If there were any streams being opened, then either
                 // the connection will be closed for an unrelated reason or a fresh reference will
                 // be constructed for the newly opened stream.
-                conn.implicit_close();
+                conn.implicit_close(&self.shared);
             }
         }
     }
 }
 
 impl std::ops::Deref for ConnectionRef {
-    type Target = Mutex<ConnectionInner>;
+    type Target = ConnectionInner;
     fn deref(&self) -> &Self::Target {
-        &self.0
+        &*self.0
     }
 }
 
+#[derive(Debug)]
 pub struct ConnectionInner {
+    pub(crate) state: Mutex<State>,
+    pub(crate) shared: Shared,
+}
+
+#[derive(Debug)]
+pub(crate) struct Shared {
+    stream_opening: [Notify; 2],
+    closed: Notify,
+}
+
+pub(crate) struct State {
```

Comment:
Maybe a more specific name? `ConnectionState`, or `ConnState`? (I think variable names can probably still be `state`, but still think this would help clarify.)

Reply:
I think it's usually a mistake to repeat the module name in a type unless that type is reexported from a higher level. It duplicates context the module itself already establishes.